### ISSUE TITLE
Support `@ExpectedToFail(onlyIf=MyException.class)`

### ISSUE THREAD

**Issue by knutwannheden** [junit-pioneer/junit-pioneer#769]
I really loved using JUnit 4's `@Test(expected)` for the use case for which `@ExpectedToFail` was designed thanks for this!). The only thing I am missing is a way to restrict this feature to only apply when the test case failed with a given exception. Sometimes I would for example throw an `UnsupportedOperationException` in the production code and only want the corresponding tests not to fail, when that is the reason, not when there is an `AssertionError` (which may instead indicate a bug).

**Comment by beatngu13** [comment]
I like that idea! With `onExceptions`, we already have something similar in `@RetryingTest`:

https://junit-pioneer.org/docs/retrying-test/#configuring-on-which-exceptions-to-retry

Not sure if I would go with `onlyIf`, but that's a detail. What does everybody else think?

If feedback is positive, @knutwannheden would you be interested in giving this a shot (i.e. provide a PR)?

**Comment by nipafx** [comment]
Quoting from the `@ExpectedToFail` documentation:

> This annotation is not intended as a way to mark test methods which intentionally cause exceptions. Such test methods should use [JUnit’s assertThrows](https://junit.org/junit5/docs/current/api/org.junit.jupiter.api/org/junit/jupiter/api/Assertions.html#assertThrows(java.lang.Class,org.junit.jupiter.api.function.Executable)) or similar means to explicitly test for a specific exception class being thrown by a specific action.

**Comment by knutwannheden** [comment]
@nipafx Yes, I get that part and I don't see any real contradiction to my use case. The current annotation can be thought to have `onExceptions=Throwable.class`.

In the use case I outlined I have developed a set of tests (e.g. to match some specification) and in the implementation I throw `UnsupportedOperationException`s. Now as I make progress I fix the production code by replacing the exception with a proper implementation, at which point the `@ExpectedToFail` annotation can be removed again.

**Comment by nipafx** [comment]
Yeah, I see your point. `@ExpectedToFail` essentially says "I know this code fails - let me know when that changes" and with your addition, it's "I know this code fails _for a specific reason_ - let me know when that changes". Makes sense, I'll have a look at the PR-
